pkg/ddc/thin/referencedataset/engine.go

```diff
@@ -0,0 +1,173 @@
+	mountedRuntimeInfo base.RuntimeInfoInterface
+}
+
+// ID returns the id of the engine
+func (e *VirtualDatasetEngine) ID() string {
+	return e.Id
+}
+
+// BuildVirtualDatasetThinEngine build engine for handling virtual dataset
+func BuildVirtualDatasetThinEngine(id string, ctx cruntime.ReconcileRequestContext) (base.Engine, error) {
+	// currently, ctx has no dataset
+	engine := &VirtualDatasetEngine{
+		Id:          id,
+		Client:      ctx.Client,
+		name:        ctx.Name,
+		namespace:   ctx.Namespace,
+		runtimeType: ctx.RuntimeType,
+	}
+	engine.Log = ctx.Log.WithValues("virtual engine", ctx.RuntimeType).WithValues("id", id)
+
+	// check if support the dataset mount format
+	err := engine.checkDatasetMountSupport()
+	if err != nil {
+		return nil, err
+	}
+
+	// Build and setup runtime info
+	_, err = engine.getRuntimeInfo()
+	if err != nil {
+		return nil, fmt.Errorf("engine %s failed to get runtime info", ctx.Name)
+	}
+
+	// set sync duration
+	duration, err := getSyncRetryDuration()
+	if err != nil {
+		engine.Log.Error(err, "Failed to parse syncRetryDurationEnv: FLUID_SYNC_RETRY_DURATION, use the default setting")
+	}
+	if duration != nil {
+		engine.syncRetryDuration = *duration
+	} else {
+		engine.syncRetryDuration = defaultSyncRetryDuration
+	}
+	engine.timeOfLastSync = time.Now().Add(-engine.syncRetryDuration)
+	engine.Log.Info("Set the syncRetryDuration", "syncRetryDuration", engine.syncRetryDuration)
+
+	// get the mountedRuntimeInfo
+	_, err = engine.getMountedRuntimeInfo()
+	if err != nil {
+		return nil, fmt.Errorf("engine %s failed to get mounted dataset's runtime info", ctx.Name)
+	}
+
+	return engine, err
+}
+
+func (e *VirtualDatasetEngine) Setup(ctx cruntime.ReconcileRequestContext) (ready bool, err error) {
+	// 1. get the physical datasets according the virtual dataset
+	dataset := ctx.Dataset
+
+	physicalNameSpacedNames := getMountedDatasetNamespacedName(dataset)
+	if len(physicalNameSpacedNames) != 1 {
+		return false, fmt.Errorf("ThinEngine with no profile name can only handle dataset only mounting one dataset")
+	}
+	namespacedName := physicalNameSpacedNames[0]
+	mountedDataset, err := utils.GetDataset(ctx.Client, namespacedName.Name, namespacedName.Namespace)
+	if err != nil {
+		return false, err
+	}
+
+	// 2. get runtime according to dataset status runtimes field
+	runtimes := mountedDataset.Status.Runtimes
+
+	if len(runtimes) == 0 {
+		return false, fmt.Errorf("mounting dataset is not bound to a runtime yet")
+	}
+
+	// 3. add this dataset to mounted dataset DatasetRef field
+	datasetRefName := getDatasetRefName(dataset.Name, dataset.Namespace)
+	if !utils.ContainsString(mountedDataset.Status.DatasetRef, datasetRefName) {
+		newDataset := mountedDataset.DeepCopy()
+		newDataset.Status.DatasetRef = append(newDataset.Status.DatasetRef, datasetRefName)
+		err := e.Client.Status().Update(context.TODO(), newDataset)
+		if err != nil {
+			return false, err
+		}
+	}
```

Comment:
Maybe it's better to use `retry.OnConflict` to retry and avoid extra reconcile loops when the update fails because of resource conflicts